fields/src/macros.rs

```diff
@@ -86,55 +86,81 @@ macro_rules! impl_primefield_from_int {
 macro_rules! sqrt_impl {
     ($Self:ident, $P:tt, $self:expr) => {{
         use crate::LegendreSymbol::*;
-        // https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5)
-        // Actually this is just normal Tonelli-Shanks; since `P::Generator`
-        // is a quadratic non-residue, `P::ROOT_OF_UNITY = P::GENERATOR ^ t`
-        // is also a quadratic non-residue (since `t` is odd).
+        // https://eprint.iacr.org/2020/1407.pdf (page 4, algorithm 1)
         match $self.legendre() {
             Zero => Some(*$self),
             QuadraticNonResidue => None,
             QuadraticResidue => {
-                let mut z = $Self::two_adic_root_of_unity();
-                let mut w = $self.pow($P::T_MINUS_ONE_DIV_TWO);
-                let mut x = w * $self;
-                let mut b = x * w;
-
-                let mut v = $P::TWO_ADICITY as usize;
-                // t = self^t
-                #[cfg(debug_assertions)]
-                {
-                    let mut check = b;
-                    for _ in 0..(v - 1) {
-                        check.square_in_place();
-                    }
-                    if !check.is_one() {
-                        panic!("Input is not a square root, but it passed the QR test")
-                    }
+                let n = $P::TWO_ADICITY;
+                // `T` is equivalent to `m` in the paper.
+                let v = $self.pow($P::T_MINUS_ONE_DIV_TWO);
+                let x = $self.pow($P::T);
+                let g = $Self::two_adic_root_of_unity().pow($P::T);
+                let two = $Self::from(2u8);
+                let n_field = $Self::from(n);
+
+                let k = ((n - 1) as f64).sqrt().floor() as usize;
+                let k_field = $Self::from(k as u64);
+                let k_2 = ((k as f64) / 2.0).floor() as usize;
+                let k_2_field = $Self::from(k_2 as u64);
+                let k_1 = k - k_2;
+                let l_minus_one_times_k = n_field - $Self::one() - k_2_field;
+                let l_minus_one = l_minus_one_times_k / k_field;
+                let l = l_minus_one + $Self::one();
+                let mut l_s: Vec<$Self> = Vec::with_capacity(k);
+
+                for _ in 0..k_1 {
+                    l_s.push(l_minus_one);
+                }
```

Comment:
```suggestion
                for _ in 0..k_1 {
                    l_s.push(l_minus_one);
                }
```
l_s.resize(l_s.len() + k_1, l_minus_one);